File: backend/app/models/github.py
import uuid
from datetime import datetime
from sqlalchemy import String, ForeignKey, DateTime, Integer, Boolean
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database.db import Base

class GithubRepository(Base):
    __tablename__ = "github_repositories"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    workspace_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("workspaces.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    full_name: Mapped[str] = mapped_column(String(255), nullable=False)
    html_url: Mapped[str] = mapped_column(String(500), nullable=False)

class GithubPullRequest(Base):
    __tablename__ = "github_pull_requests"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    repository_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("github_repositories.id", ondelete="CASCADE"), nullable=False)
    number: Mapped[int] = mapped_column(Integer, nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    state: Mapped[str] = mapped_column(String(50), default="open", nullable=False) # open, closed, merged
    html_url: Mapped[str] = mapped_column(String(500), nullable=False)
    task_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("tasks.id", ondelete="SET NULL"), nullable=True)

    task = relationship("Task", backref="pull_requests")

class GithubCommit(Base):
    __tablename__ = "github_commits"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    repository_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("github_repositories.id", ondelete="CASCADE"), nullable=False)
    sha: Mapped[str] = mapped_column(String(100), nullable=False)
    message: Mapped[str] = mapped_column(String(500), nullable=False)
    author: Mapped[str] = mapped_column(String(100), nullable=False)
    html_url: Mapped[str] = mapped_column(String(500), nullable=False)

class GithubIssue(Base):
    __tablename__ = "github_issues"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    repository_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("github_repositories.id", ondelete="CASCADE"), nullable=False)
    number: Mapped[int] = mapped_column(Integer, nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    state: Mapped[str] = mapped_column(String(50), default="open", nullable=False)
    html_url: Mapped[str] = mapped_column(String(500), nullable=False)

class GithubBranch(Base):
    __tablename__ = "github_branches"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    repository_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("github_repositories.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    protected: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

class GithubDeployment(Base):
    __tablename__ = "github_deployments"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    repository_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("github_repositories.id", ondelete="CASCADE"), nullable=False)
    environment: Mapped[str] = mapped_column(String(100), nullable=False)
    status: Mapped[str] = mapped_column(String(50), default="pending", nullable=False) # success, failure, pending
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

class GithubWorkflowRun(Base):
    __tablename__ = "github_workflow_runs"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    repository_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("github_repositories.id", ondelete="CASCADE"), nullable=False)
    run_number: Mapped[int] = mapped_column(Integer, nullable=False)
    event: Mapped[str] = mapped_column(String(50), nullable=False) # push, pull_request
    status: Mapped[str] = mapped_column(String(50), nullable=False) # completed, in_progress, queued
    conclusion: Mapped[str | None] = mapped_column(String(50), nullable=True) # success, failure, cancelled
    html_url: Mapped[str] = mapped_column(String(500), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)

File: frontend/services/github.ts
import { api } from './api'

export interface GithubRepository {
  id: string
  workspace_id: string
  name: string
  full_name: string
  html_url: string
}

export interface GithubPullRequest {
  id: string
  repository_id: string
  number: number
  title: string
  state: string
  html_url: string
  task_id?: string
}

export interface GithubCommit {
  id: string
  repository_id: string
  sha: string
  message: string
  author: string
  html_url: string
}

export interface GithubIssue {
  id: string
  repository_id: string
  number: number
  title: string
  state: string
  html_url: string
}

export interface GithubBranch {
  id: string
  repository_id: string
  name: string
  protected: boolean
}

export interface GithubDeployment {
  id: string
  repository_id: string
  environment: string
  status: string
  updated_at: string
}

export const githubService = {
  async getOAuthUrl(): Promise<string> {
    const response = await api.get<{ url: string }>('/github/oauth/url')
    return response.data.url
  },

  async callbackOAuth(code: string): Promise<{ status: string; username: string }> {
    const response = await api.post('/github/oauth/callback', { code })
    return response.data
  },

  async getRepositories(): Promise<GithubRepository[]> {
    const response = await api.get<GithubRepository[]>('/github/repositories')
    return response.data
  },

  async getPullRequests(repositoryId?: string): Promise<GithubPullRequest[]> {
    const response = await api.get<GithubPullRequest[]>('/github/pull-requests', {
      params: repositoryId ? { repository_id: repositoryId } : {}
    })
    return response.data
  },

  async linkTaskToPR(prId: string, taskId: string): Promise<GithubPullRequest> {
    const response = await api.patch<GithubPullRequest>(`/github/pull-requests/${prId}/link-task`, {
      task_id: taskId
    })
    return response.data
  },

  async getCommits(repositoryId: string): Promise<GithubCommit[]> {
    const response = await api.get<GithubCommit[]>('/github/commits', {
      params: { repository_id: repositoryId }
    })
    return response.data
  },

  async getIssues(repositoryId: string): Promise<GithubIssue[]> {
    const response = await api.get<GithubIssue[]>('/github/issues', {
      params: { repository_id: repositoryId }
    })
    return response.data
  },

  async getBranches(repositoryId: string): Promise<GithubBranch[]> {
    const response = await api.get<GithubBranch[]>('/github/branches', {
      params: { repository_id: repositoryId }
    })
    return response.data
  },

  async getDeployments(repositoryId: string): Promise<GithubDeployment[]> {
    const response = await api.get<GithubDeployment[]>('/github/deployments', {
      params: { repository_id: repositoryId }
    })
    return response.data
  }
}
